doc/developer-guide/architecture/threads-and-events.en.rst

```diff
@@ -0,0 +1,135 @@
+
+Despite the name :class:`EventProcessor` is primarily a thread management class. It enables the creation and management of thread groups which are then used by the |TS| core for different types of computation. The set of groups is determined at run time via the :func:`EventProcessor::registerEventType`. Threads managed by :class:`EventProcessor` have the :class:`EThread` start continuation controlled by :class:`EventProcessor`. Instead each thread group (event type) has a list of continuations to run when a thread of that type starts. Continuations are added to the list with :func:`EventProcessor::schedule_spawn`. There are two variants of this method, one for continuations and one for just a function. The latter creates a continuation to call the function and then schedules that using the former.
+
+:class:`EventProcessor` is intended to be a singleton and the global instance is :var:`eventProcessor`.
+
+In general if a subsystem in the |TS| core is setting up a thread group, it should use code of the form ::
+
+   int ET_GROUP; // global variable, where "GROUP" is repalced by the actual group / type name.
+   int n_group_threads = 3; // Want 3 of these threads by default, possibly changed by configuration options.
+   constexpr size_t GROUP_STACK_SIZE = 1 << 20; // stack size for each thread.
+   void Group_Thread_Init(EThread*); // function to perform per thread local initialization.
+   //
+   ET_GROUP = eventProcessor::registerEventType("Group");
+   eventProcessor.schedule_spawn(&Group_Per_Thread_Init, ET_GROUP);
+   eventProcessor.spawn_event_threads(ET_GROUP, n_group_threads, GROUP_STACK_SIZE);
+
+
+The function :code:`Group_Thread_Init` can be replaced with a continuation if that's more convenient. One advantage of a continuation is additional data (via :arg:`cookie`) can be provide during thread initialization.
+
+If there is no thread initializatoin needed, this can be compressed in to a single call ::
+
+   ET_GROUP = eventProcessor.spawn_event_threads("Group", n_group_threads, GROUP_STACK_SIZE);
+
+This registers the group name and type, starts the threads, and returns the event type.
+
+TS-4265 Changes
+===============
+
+The essential change for TS-4625 is to provide a mechanism to arrange for code to be executed on :class:`EventProcessor` threads before starting event processing. This enables robust initialization of thread local data / state. Previously either the initialization would be done after thread start from a different thread by iterating over the threads, or "foreign" code being embdedded in the low level thread / event processing logic.
+
+Types
+=====
+
+.. type:: EventType
+
+   A thread classification value that represents the type of events the thread is expected to process.
+
+.. var:: EventType ET_CALL
+
+   A predefined :type:`EventType` which always exists. This is deprecated, use :var:`ET_NET` instead.
```

Comment:
It is deprecated? That ought to be in a code comment or compiler annotation if that's so.